### PR TITLE
Introduce new "help" command

### PR DESCRIPTION
**Pull request by prakashsurya**
This change adds a new "help" command that can be used to pring out
useful "help messages" for any particular SDB command. The contents of
the help message is derived automatically from the command class's
docstring and argument parser.

Co-authored-by: Prakash Surya <prakash.surya@delphix.com>

### CODE REVIEW

**Comment by prakashsurya** [comment]
cc: @sdimitro @ctuffli @ahrens 

This is a continuation/extention of: https://github.com/delphix/sdb/pull/68 .. I've taken what @ctuffli originally proposed, and tried to get it into a state that can be merged. This PR has the following differences from what was originally proposed:

- I've removed the `-v` flag to the `help` command; now it always emits verbose output
- I've removed the ability to run `help` without specifying a command; I think we should revisit this in a follow up PR/bug, to determine how to address the "what commands exist?" question raised in the original PR
- I've integrated with each command's parser, so we get all of the argparse information in the help message (e.g. description, options, etc)
- I refactored how the parser object gets created, so that we can create it in a `@classmethod` without instantiating a new object
- I removed the changes that add new docstrings to the commands; I'd like to do this in a seperate commit

As is, the format of the help message is the following:
```
> help spa
SUMMARY
    spa [-h] [-v] [-m] [-H] [-w] [poolnames [poolnames ...]]

    positional arguments:
      poolnames

    optional arguments:
      -h, --help       show this help message and exit
      -v, --vdevs      vdevs flag
      -m, --metaslab   metaslab flag
      -H, --histogram  histogram flag
      -w, --weight     weight flag

```
```
> help pp
SUMMARY
    pp [-h]

    optional arguments:
      -h, --help  show this help message and exit

ALIASES
    pretty_print, pp

```

Once we add docstrings to certain commands, I expect this to expand into something like this:
```
> help head                                                                                                          
SUMMARY                                                                                                       
    head [-h] [count]                                                                                            
                                                                                                                 
    Return the first N objects passed in the pipeline                                                                 
                                                                                                                       
    positional arguments:                                                                                   
      count                                                                                                              
                                                                                                                         
    optional arguments:                                                                                               
      -h, --help  show this help message and exit                                                                    
                                                                                                                       
DESCRIPTION                                                                                                 
                                                                                                                    
    Command takes a 'count' parameter which defaults to 10                                                       
                                                                                                                         
EXAMPLES                                                                                                             
                                                                                                                      
    > spa | head 1 | member spa_name                                                                                  
    (char [256])"domain0"  

```

Please let me know what you all think about what I have here. If we're in agreement with this approach, I think we can land this, and the circle back to the following items:

- Add new docstrings to each command
- Determine how to answer the "what commands are available?" question

File: sdb/command.py [hunk 1]
```diff
@@ -34,22 +34,68 @@ class Command:
...
+        #
+        # If the class doesn't have a docstring, "inspect.getdoc" will
+        # pull from the parent class's docstring, if the parent class
+        # does have a docstring. This is not the behavior we want, so we
+        # first check "cls.__doc__" before using "inspect.getdoc".
+        #
+        if cls.__doc__:
+            summary = inspect.getdoc(cls).splitlines()[0].strip()
+        else:
+            summary = None
```
Comment (sdimitro) [inline sdb/command.py]: Nice!

File: sdb/command.py [hunk 2]
```diff
@@ -34,22 +34,68 @@ class Command:
...
 
-    #
-    # name:
-    #    The name used when the command was invoked. This
-    #    is generally used as a parameter passed to
-    #    exceptions raised by SDB to make error messages
-    #    more precise.
-    #
-    name: str = ""
-
```
Comment (sdimitro) [inline sdb/command.py]: Out of curiosity, why is this removed? Asking because I liked the comment making the differentiation between `name` and `names`.

File: sdb/commands/address.py [hunk 3]
```diff
@@ -42,9 +42,11 @@ class Address(sdb.Command):
 
     names = ["address", "addr"]
 
-    def _init_argparse(self, parser: argparse.ArgumentParser) -> None:
-        super()._init_argparse(parser)
+    @classmethod
+    def _init_parser(cls, name: str) -> argparse.ArgumentParser:
+        parser = super(Address, cls)._init_parser(name)
```
Comment (sdimitro) [inline sdb/commands/address.py]: It's a bit annoying that we have to do it like that, is there a reason why the following doesn't work:
```
parser = super()._init_parser(name)
```
Also if you actually gave the above a try is there any advantage on using the more verbose version that you have?

File: sdb/command.py [hunk 4]
```diff
@@ -34,22 +34,68 @@ class Command:
...
+        print("SUMMARY")
+        for i, line in enumerate(parser.format_help().split('\n')):
+            #
+            # When printing the help message, the first line will have a
+            # "usage: " prefix string which looks awkward, so we strip
+            # that prefix prior to printing the first line.
+            #
+            if i == 0:
+                line = line.replace('usage: ', '')
+            print("    {}".format(line.replace('usage: ', '')))
```
[the remark this answers is not in the document]
Comment (prakashsurya) [reply sdb/command.py]: Sorry, good catch. I meant to remove the second replace (which I used first, when iterating).

While perhaps we could remove it on every line, I wanted to be more targeted and only remove it on the first line.

I'll fix this.

File: sdb/commands/address.py (hunk 3 above)
Comment (prakashsurya) [reply sdb/commands/address.py]: I can try that; I can't recall if I had tried that or not. This is a class method, though, so I think we need to do it this way. But I'll double check.. I also would much rather just use `super()` if possible.

File: sdb/command.py (hunk 2 above)
Comment (prakashsurya) [reply sdb/command.py]: So, based on my understanding of python, this will create a class level variable named `name`, which is different than the object level variable named `name`. I didn't see where we used the class level `name` variable, vs. how we do use the object level `self.name` variable.

Assuming I'm correct and this is dead code, I figured it should be removed to avoid confusion.

Let me know if I'm understanding this incorrectly.

File: sdb/commands/address.py (hunk 3 above)
Comment (prakashsurya) [reply sdb/commands/address.py]: It looks like `super()` does work, so I don't know why I didn't try that to begin with. I guess my lack of experience with Python is showing :)

File: sdb/command.py (hunk 2 above)
Comment (sdimitro) [reply sdb/command.py]: I see. I honestly had no idea that this was a thing. I'm resolving this.

**Comment by prakashsurya** [comment]
Using the github UI to squash and merge unfortunately lost the `Author` attribution to @ctuffli:
```
commit ca9d5c2d668f9bff66593d75ee351a24ce28bfc0
Author: Prakash Surya <prakash.surya@delphix.com>
Date:   Fri Nov 15 12:50:46 2019 -0800

    Introduce new "help" command (#70)
    
    This change adds a new "help" command that can be used to print out
    useful "help messages" for any particular SDB command. The contents of
    the help message is derived automatically from the command class's
    docstring and argument parser.
    
    Co-authored-by: Prakash Surya <prakash.surya@delphix.com>
```
@sdimitro should we attempt to rectify this, or just leave it as is? I'm not sure we can do anything, without a force push, and changing the commit history of the repository...

**Comment by ctuffli** [comment]
Don't sweat the attribution on this
